### ISSUE TITLE
Support training arguments for training stages

### ISSUE THREAD

**Issue by eu9ene** [hplt-project/OpusTrainer#44]
After switching from pre-training and fine-tuning tasks to OpusTrainer with multiple stages we ran into issue with early-stopping. We usually want to train early stages for a fixed number of epochs and then train the last stage until the metrics are not improving and it's stopped by Marian. This Marian parameter and probably some others should be configurable per stage. I assume we could add them to the training config:

```
start:
  mix:
  - clean 0.8
  - medium 0.2
  - dirty 0
  - until clean 2 # Until two epochs of clean
  modifiers:
    - UpperCase: 0.05
    - TitleCase: 0.05
  arguments:
    - early-stopping: 30
```

**Comment by jelmervdl** [comment]
Oh this is quite an obvious oversight in the design of OpusTrainer :(

I implemented something like this as an experiment earlier this year, where I would run marian till it would stall and only then advance to the next stage. The design for this would be similar, except not running marian till stall, but till epochs run out, and then restart with (possibly) different parameters for the next stage.

That previous work is here: https://github.com/hplt-project/OpusTrainer/compare/main...stages-till-stall

I think I can rebase and merge that, and add some sort of arguments option. It will be raw arguments, just a list that gets appended to `args.trainer`, but that should be sufficient for your use case.

**Comment by eu9ene** [comment]
That would be awesome! This is something we struggle with a lot after migration.

**Comment by jelmervdl** [comment]
@eu9ene Have a look at #45 and see whether that is the way to go. It's quick and dirty and noticed some issues I have with it in the pull request, your ideas/comments on it are more than welcome!

**Comment by XapaJIaMnu** [comment]
@eu9ene could please explain the usecase a bit more?
From the point of view of OpusTrainer, it shouldn't "know" about what is the downstream task and I feel like providing specific arguments to the downstream executable at each stage is somewhat of an overcomplication.

You can achieve what you are looking for by having two consecutive steps of your pipeline with opuscleaner but specifying different command (`runme.sh`?). OpusCleaner will reuse the state and resume at whatever stage it is in the specific dataset.

More specifically, if you are running into stalls that require an early stopping parameter of 30, it looks to me like one of the following is happening:

1) The neural network is too small for the amount of data that you have (solution: Bigger neural network).
2) The validation data is not representative of the training data but it's still high quality. (In this case, you expect the validation loss to wiggle but not drop drammatically. Nothing like 30 consecutive stalls).
3) The training data is bad and there is no point of continuing training further.

Most likely 1) or 3) is happening at which point training should stop (or advance to a new stage where more data is available).

I would say that if this is happening the correct solution is to advance the training data scheduling by one step and for that to happen automatically, OpusTrainer needs to be able to receive feedback from the child process (which we planned to do but never actually got to implement).

**Comment by eu9ene** [comment]
We ran into this [issue](https://github.com/mozilla/firefox-translations-training/issues/314) that we initially thought is related to training stages. It turned out it was a bug in hyperparameters (optimizer-delay). 

We specifically merged two different steps into one to reduce complexity and have more control on how we use back-translated data. The idea was that if back-translated data has worse quality for lower resource languages we might benefit from using different hyperparameters. In this case it would be a work around for another issue.

Now when the bug is fixed, the training curve looks pretty normal and it trains fine with early-stopping: 20, so this feature is optional for us.  It might indeed be an overcomplication (we discussed related issues with @jelmervdl  in #45) if the model trains fine on different stages with the same parameters. We need to run more experiments though. If in your experience there are no issues with this approach we'll try to follow your lead and solve issues with other means if they arise. It's an interesting idea to propagate feedback to OpusTrainer. I'm not sure how it would work though. 

Feel free to close this issue. We can get back to this when we run into another problem that's hard to fix with other means.
